clang/lib/Sema/SemaDecl.cpp

```diff
@@ -14690,6 +14690,17 @@ void Sema::FinalizeDeclaration(Decl *ThisDecl) {
     }
   }
 
+  // The result of __builtin_counted_by_ref cannot be assigned to a variable.
+  // It allows leaking and modification of bounds safety information.
+  if (const auto *CE = dyn_cast_if_present<CallExpr>(VD->getInit())) {
+    const Expr *E = CE->getCallee()->IgnoreParenImpCasts();
+    if (const BuiltinType *PTy = E->getType()->getAsPlaceholderType();
+        PTy && PTy->getKind() == BuiltinType::BuiltinCountedByRef)
+      Diag(E->getExprLoc(),
+           diag::err_builtin_counted_by_ref_cannot_leak_reference)
+          << E->getSourceRange();
+  }
```

Comment:
```suggestion
  if (const BuiltinType *PTy = E->getType()->getAsPlaceholderType();
      PTy && PTy->getKind() == BuiltinType::BuiltinCountedByRef) {
    Diag(E->getExprLoc(),
         diag::err_builtin_counted_by_ref_cannot_leak_reference)
        << E->getSourceRange();
  }
```
I think we should just check for the placeholder type instead of for a call expression here. 

Also, even though I’m someone who generally prefers not to use braces around the body of an `if` when possible, I think a 5-line `if` statement w/o braces is beyond the limits of what I’d consider to be legible. ;Þ

One more thing, I feel like `AddInitializerToDecl` might be a better place to put this (maybe right before the `if (VDecl->getType()->isUndeducedType())` part?).

Reply:
Also, I find it a bit weird that we never check for placeholders here—weird but possible, that is, because initialisers are generally also a bit weird... can you try removing this check entirely?

Reply:
If I remove the check then it won't trigger an error. Initialization and assignment are two completely different paths.